core/src/cluster_slot_state_verifier.rs

```diff
@@ -0,0 +1,610 @@
+use crate::{
+    fork_choice::ForkChoice, heaviest_subtree_fork_choice::HeaviestSubtreeForkChoice,
+    progress_map::ProgressMap,
+};
+use solana_sdk::{clock::Slot, hash::Hash};
+use std::collections::{BTreeMap, HashMap, HashSet};
+
+pub type GossipConfirmedSlots = BTreeMap<Slot, Hash>;
+type SlotStateHandler = fn(Slot, &Hash, Option<&Hash>, bool, bool) -> Vec<ResultingStateChange>;
+
+#[derive(PartialEq, Debug)]
+pub enum SlotStateUpdate {
+    Frozen,
+    Confirmed,
+    Dead,
+    Duplicate,
+}
+
+#[derive(PartialEq, Debug)]
+pub enum ResultingStateChange {
+    MarkSlotDuplicate,
+    RepairConfirmedVersion(Hash),
+    ConfirmedSlotMatchesCluster,
+}
+
+impl SlotStateUpdate {
+    fn to_handler(&self) -> SlotStateHandler {
+        match self {
+            SlotStateUpdate::Dead => on_dead_slot,
+            SlotStateUpdate::Frozen => on_frozen_slot,
+            SlotStateUpdate::Confirmed => on_cluster_update,
+            SlotStateUpdate::Duplicate => on_cluster_update,
+        }
+    }
+}
+
+fn repair_correct_version(_slot: Slot, _hash: &Hash) {}
+
+fn on_dead_slot(
+    slot: Slot,
+    bank_frozen_hash: &Hash,
+    cluster_confirmed_hash: Option<&Hash>,
+    _is_slot_duplicate: bool,
+    is_dead: bool,
+) -> Vec<ResultingStateChange> {
+    assert!(is_dead);
+    // Bank should not have been frozen if the slot was marked dead
+    assert_eq!(*bank_frozen_hash, Hash::default());
+    if let Some(cluster_confirmed_hash) = cluster_confirmed_hash {
+        // If the cluster confirmed some version of this slot, then
+        // there's another version
+        warn!(
+            "Cluster confirmed slot {} with hash {}, but we marked slot dead",
+            slot, cluster_confirmed_hash
+        );
+        // No need to check `is_slot_duplicate` and modify fork choice as dead slots
+        // are never frozen, and thus never added to fork choice. The state change for
+        // `MarkSlotDuplicate` will try to modify fork choice, but won't find the slot
+        // in the fork choice tree, so is equivalent to a
+        return vec![
+            ResultingStateChange::MarkSlotDuplicate,
+            ResultingStateChange::RepairConfirmedVersion(*cluster_confirmed_hash),
+        ];
+    }
+
+    vec![]
+}
+
+fn on_frozen_slot(
+    slot: Slot,
+    bank_frozen_hash: &Hash,
+    cluster_confirmed_hash: Option<&Hash>,
+    is_slot_duplicate: bool,
+    is_dead: bool,
+) -> Vec<ResultingStateChange> {
+    // If a slot is marked frozen, the bank hash should not be default,
+    // and the slot should not be dead
+    assert!(*bank_frozen_hash != Hash::default());
+    assert!(!is_dead);
+
+    if let Some(cluster_confirmed_hash) = cluster_confirmed_hash {
```

Comment:
@behzadnouri this is where I imagine we'll handle the case where we detect the cluster has confirmed some version of the hash, after detection is piped in from here: https://github.com/solana-labs/solana/pull/9698/files#r545028003